engine/function/src/templates/Numeric.ftl

```diff
@@ -1268,38 +1268,63 @@ public class Numeric {
 
         if (n == 0) {
             return NULL_DOUBLE;
-        } else {
-            // NULL values sorted to beginning of the array.
-            ${pt.primitive}[] copy = sort(values.toArray());
+        }
+
+        ${pt.primitive}[] sorted = values.copyToArray();
+        Arrays.sort(sorted);
 
-            // Determine if there are any NULL in the array.
-            int nullCount = 0;
-            for (int i = 0; i < n; i++) {
-                if (isNull(copy[i])) {
+        int nullStart = -1;
+        int nullCount = 0;
+
+        <#if pt.valueType.isFloat >
+        for (int i = 0; i < n; i++) {
```

Comment:
This can be better.
1. If we have seen any value that compares greater than `QueryConstants.NULL_X`, we can short-circuit the forward iteration. There will be no nulls.
2. The NaN search can be from the back, because NaNs sort after other values, and similarly can short-circuit.